aws-auth-cognito/src/test/java/com/amplifyframework/auth/cognito/StateTransitionTests.kt

```diff
@@ -300,15 +308,106 @@ class StateTransitionTests : StateTransitionTestBase() {
-                if (authNState is AuthenticationState.SignedIn && authZState is AuthorizationState.SessionEstablished) {
+        assertTrue { testLatch.await(5, TimeUnit.SECONDS) }
+        assertTrue { configureLatch.await(5, TimeUnit.SECONDS) }
+    }
+
+    @Test
+    fun testSignInWithCustom() {
+        setupConfigureSignedOut()
+        setupSignInActionWithCustomAuth()
+        setupCustomAuthActions()
+
+        Mockito.`when`(mockAuthenticationActions.initiateSignInAction(MockitoHelper.anyObject()))
+            .thenReturn(
+                Action { dispatcher, _ ->
+                    dispatcher.send(
+                        SignInEvent(
+                            SignInEvent.EventType.InitiateSignInWithCustom(
+                                "username",
+                                "password",
+                                AWSCognitoAuthSignInOptions.builder().authFlowType(AuthFlowType.CUSTOM_AUTH).metadata(
+                                    mapOf()
+                                ).build()
+                            )
+                        )
+                    )
+                }
+            )
+
+        Mockito.`when`(signedInData.cognitoUserPoolTokens)
+            .thenReturn(
+                CognitoUserPoolTokens("", "", "", 0)
+            )
+
+        val testLatch = CountDownLatch(1)
+        val configureLatch = CountDownLatch(1)
+        val subscribeLatch = CountDownLatch(1)
+        var token: StateChangeListenerToken? = null
+        token = stateMachine.listen(
+            {
+                val authState =
+                    it.takeIf { it is AuthState.Configured && it.authNState is AuthenticationState.SignedOut }
+                authState?.run {
+                    configureLatch.countDown()
+                    stateMachine.send(
+                        AuthenticationEvent(
+                            AuthenticationEvent.EventType.SignInRequested(
+                                "username",
+                                "password",
+                                AWSCognitoAuthSignInOptions.builder()
+                                    .metadata(mapOf())
+                                    .authFlowType(AuthFlowType.CUSTOM_AUTH)
+                                    .build()
+                            )
+                        )
+                    )
+                }
+
+                val challengeState = it.authNState?.signInState?.challengeState.takeIf {
+                    signInChallengeState ->
+                    signInChallengeState is SignInChallengeState.WaitingForAnswer
+                }
+                challengeState?.apply {
+                    stateMachine.send(
+                        SignInChallengeEvent(
+                            SignInChallengeEvent.EventType.VerifyChallengeAnswer("test", mapOf())
+                        )
+                    )
+                }
+
+                val authNState =
```

Comment:
`authNState is AuthenticationState.SignedIn && authZState is AuthorizationState.SessionEstablished` represents the correct signed in state.